mart/callbacks/adversarial_training.py

```diff
@@ -0,0 +1,46 @@
+#
+# Copyright (C) 2022 Intel Corporation
+#
+# SPDX-License-Identifier: BSD-3-Clause
+#
+
+from pytorch_lightning.callbacks import Callback
+
+from mart.models import LitModular
+
+__all__ = ["AdversarialTraining"]
+
+
+class AdversarialTraining(Callback):
+    """Perturbs inputs to be adversarial."""
+
+    def __init__(
+        self, adversary=None, train_adversary=None, validation_adversary=None, test_adversary=None
+    ):
+        adversary = adversary or train_adversary
+
+        self.train_adversary = train_adversary or adversary
+        self.validation_adversary = validation_adversary or adversary
+        self.test_adversary = test_adversary or adversary
+
+    def on_train_batch_start(self, trainer, pl_module, batch, batch_idx):
+        input, target = batch
+
+        # FIXME: We reach into LitModular here...how can we get rid of this?
+        assert isinstance(pl_module, LitModular)
+        model = pl_module.model
+        sequence = model._sequences["training"]
+
+        # FIXME: This doesn't work because sequence does not include the Adversary module. How can we fix that?
+        #        Because this a callback, we can safely assume the Adversary module should live before the model.
+        #        We should be able to "manually" insert it into the sequence here.
```

Comment:
Maybe we should make `mart.attack.Adversary` agnostic to the underlying model? That is, `Adversary` currently assumes the model it is attacking is a `LitModular`. Can we fix that?

Reply:
It feels like there should be an `AdversaryLayer` (that is a `Module`) that wraps (or extends) an `Adversary` (that is `LightningModule`).